File: Final_Project/Code/Project_4/Bai_toan_5/Problems/6/CSS_bt5_problem6.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Tree {
public:
    int n;                       // so dinh
    vector<vector<int>> adj;     // danh sach ke

    Tree(int vertices) : n(vertices) {
        adj.resize(n);
    }

    void add_edge(int u, int v) {
        // Them canh (u,v)
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    bool is_tree() {
        // Kiem tra xem do thi co phai la cay
        // visited: trang thai tham
        // parent: cha cua moi dinh trong BFS
        int edge_count = 0;
        for (int v = 0; v < n; ++v) edge_count += adj[v].size();
        if (edge_count / 2 != n - 1) return false;

        vector<bool> visited(n, false);
        vector<int> parent(n, -1);
        queue<int> q;
        q.push(0);
        visited[0] = true;

        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int v : adj[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    parent[v] = u;
                    q.push(v);
                } else if (v != parent[u]) {
                    return false;
                }
            }
        }
        for (bool v : visited) if (!v) return false;
        return true;
    }
};

int main() {
    Tree t(4);
    t.add_edge(0, 1);
    t.add_edge(0, 2);
    t.add_edge(2, 3);
    cout << "Is tree: " << t.is_tree() << endl; // 1
    return 0;
}

File: Final_Project/Code/Project_4/Bai_toan_5/Problems/6/CSS_bt5_problem6.py
from collections import deque

class Tree:
    def __init__(self, n):
        # n: so dinh
        # adj: danh sach ke, dai dien cho cac canh
        self.n = n
        self.adj = [[] for _ in range(n)]

    def add_edge(self, u, v):
        # Them canh (u,v)
        self.adj[u].append(v)
        self.adj[v].append(u)

    def is_tree(self):
        # Kiem tra xem do thi co phai la cay
        # visited: trang thai tham
        # parent: cha cua moi dinh trong BFS
        if sum(len(edges) for edges in self.adj) // 2 != self.n - 1:
            return False
        visited = [False] * self.n
        parent = [-1] * self.n
        queue = deque([0])
        visited[0] = True

        while queue:
            u = queue.popleft()
            for v in self.adj[u]:
                if not visited[v]:
                    visited[v] = True
                    parent[v] = u
                    queue.append(v)
                elif v != parent[u]:
                    return False
        return all(visited)

# Vi du su dung
if __name__ == "__main__":
    t = Tree(4)
    t.add_edge(0, 1)
    t.add_edge(0, 2)
    t.add_edge(2, 3)
    print("Is tree:", t.is_tree())  # True